### PR TITLE
Options downloading

### PR DESCRIPTION
**Pull request by CremaLuca**
## Premise
Even though Professor Peserico [told us not to](https://www.youtube.com/watch?v=CxeWlx2p2tw) download options because we don't have the power yet, I did it anyway because I was bored.

#### feat `otri.downloader.option_downloader`
An abstract class that contains the functionalities that YahooFinance gives for options, I assumed most of the other data sources could have the same functionalities:
 - List of expiration dates
 - Get current option chains for puts and calls
 - Get a history of last trades for every contract for puts and calls
 - Get name/ticker of every contract in puts and calls

#### feat `otri.downloader.yahoo_downloader`
I didn't want to create another file so I changed the name of the yahoo downloader and added a new class for option downloading.
Implements `option_downloader` and re-uses the time-series downloader to download historical data about contracts (given its ticker).

#### fix autocorrelation:
Changed names according to filename and classname changes

#### feat `option_download.py`
Script to download options data: downloads puts chains, calls chains, and for every contract it downloads history data.
It works with multithreading to speed things up: with 55 threads it takes 19 hours to download all of russell3000 data (635MB)

#### feat `logger`
Variable level of logging into the console.

#### feat `key_handler`
Now has apply deep to values.

## Picture of a cute animal
Cute

### CODE REVIEW

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- option_download.py  10
- otri/downloader/options_downloader.py  1
         

Clones added
============
- timeseries_download.py  3
- option_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963&bid=19342108)

**Review by ricdezen — APPROVED** [review]

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- otri/downloader/options_downloader.py  1
- option_download.py  10
         

Clones added
============
- timeseries_download.py  3
- option_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963&bid=19342108)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- option_download.py  10
- otri/downloader/options_downloader.py  1
         

Clones added
============
- option_download.py  3
- timeseries_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963&bid=19342108)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- option_download.py  10
- otri/downloader/options_downloader.py  1
         

Clones added
============
- timeseries_download.py  3
- option_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963&bid=19342108)

File: otri/downloader/yahoo_downloader.py [hunk 1]
```diff
@@ -1,58 +1,63 @@
+from typing import Union, Sequence
 from datetime import date, datetime
-from .timeseries_downloader import TimeseriesDownloader, METADATA_KEY, META_INTERVAL_KEY, META_PROVIDER_KEY, META_TICKER_KEY, ATOMS_KEY, Union
+from .timeseries_downloader import TimeseriesDownloader, METADATA_KEY, META_INTERVAL_KEY, META_PROVIDER_KEY, META_TICKER_KEY, META_TYPE_KEY, META_TYPE_VALUE, ATOMS_KEY
+from .options_downloader import OptionsDownloader, META_TYPE_KEY, META_DOWNLOAD_TIME, META_EXPIRATION_DATE
```
Comment (CremaLuca) [inline otri/downloader/yahoo_downloader.py:4]: Codacy found an issue: [Reimport 'META_TYPE_KEY' (imported line 3)](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- otri/downloader/options_downloader.py  1
- option_download.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963&bid=19342108)

File: option_download.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    else:
+        log.e("{}: option_download.py -p <provider: {}> -f <ticker file: {}>".format(
+            msg,
+            list(DOWNLOADERS.keys()),
+            list_tickers_file(TICKER_LISTS_FOLDER)
+        )
+        )
+
+
+if __name__ == "__main__":
```
Comment (CremaLuca) [inline option_download.py:143]: Codacy found an issue: [If 145 is too complex (16) (MC0001)](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5905963)

**Pull request merged**